src/main/java/com/google/cloud/spanner/pgadapter/statements/CopyStatement.java

```diff
@@ -0,0 +1,267 @@
+import com.google.cloud.spanner.pgadapter.parsers.copy.CopyTreeParser;
+import com.google.cloud.spanner.pgadapter.utils.MutationBuilder;
+import com.google.spanner.v1.TypeCode;
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.csv.CSVFormat;
+
+public class CopyStatement extends IntermediateStatement {
+
+  private static final String COLUMN_NAME = "column_name";
+  private static final String SPANNER_TYPE = "spanner_type";
+  private static final String CSV = "CSV";
+
+  protected String tableName;
+  protected List<String> copyColumnNames;
+  protected String formatType;
+  protected String nullString;
+  protected char delimiterChar;
+  protected char quoteChar;
+  protected char escapeChar;
+  protected boolean hasHeader;
+  protected int formatCode;
+
+  protected Map<String, TypeCode> tableColumns;
+  protected MutationBuilder mutationBuilder;
+
+  public CopyStatement(String sql, Connection connection) throws SQLException {
+    super();
+    this.sql = sql;
+    this.command = parseCommand(sql);
+    this.connection = connection;
+    this.statement = connection.createStatement();
+  }
+
+  /** @return Mapping of table column names to column type. */
+  public Map<String, TypeCode> getTableColumns() {
+    return this.tableColumns;
+  }
+
+  protected void setTableColumns(Map tableColumns) {
+    this.tableColumns = tableColumns;
+  }
+
+  /** @return CSVFormat for parsing copy data based on COPY statement options specified. */
+  public CSVFormat getParserFormat() {
+    CSVFormat format = CSVFormat.POSTGRESQL_TEXT;
+    if (this.formatType != null && this.formatType.equalsIgnoreCase("CSV")) {
+      format = CSVFormat.POSTGRESQL_CSV;
+    }
+    if (this.nullString != null && !this.nullString.isEmpty()) {
+      format = format.withNullString(this.nullString);
+    }
+    if (this.delimiterChar != '\0') {
+      format = format.withDelimiter(this.delimiterChar);
+    }
+    if (this.escapeChar != '\0') {
+      format = format.withEscape(this.escapeChar);
+    }
+    if (this.quoteChar != '\0') {
+      format = format.withQuote(this.quoteChar);
+    }
+    format = format.withHeader(this.tableColumns.keySet().toArray(new String[0]));
+    return format;
+  }
+
+  public String getTableName() {
+    return this.tableName;
+  }
+
+  protected void setTableName(String tableName) {
+    this.tableName = tableName;
+  }
+
+  /** @return List of column names specified in COPY statement, if provided. */
+  public List<String> getCopyColumnNames() {
+    return this.copyColumnNames;
+  }
+
+  protected void setCopyColumnNames(List<String> copyColumnNames) {
+    this.copyColumnNames = copyColumnNames;
+  }
+
+  /** @return Format type specified in COPY statement, if provided. */
+  public String getFormatType() {
+    return this.formatType;
+  }
+
+  protected void setFormatType(String formatType) {
```

Comment:
I removed all of the setters and made them private instead of protected.  We can't make them final because many of them are set when execute() is called and the user supplied statement is parsed.